zarr/v3/group.py

```diff
@@ -1,97 +1,108 @@
 from __future__ import annotations
 
+import asyncio
 import json
-from typing import Any, Dict, Literal, Optional, Union
+import logging
+from typing import Any, Dict, Literal, Optional, Union, AsyncIterator, Iterator, List
 
-from attr import asdict, evolve, field, frozen
+from attr import asdict, field, frozen  # , validators
 
-from zarr.v3.array import Array
-from zarr.v3.common import ZARR_JSON, make_cattr
-from zarr.v3.metadata import RuntimeConfiguration
+from zarr.v3.array import AsyncArray, Array
+from zarr.v3.attributes import Attributes
+from zarr.v3.common import ZARR_JSON, ZARRAY_JSON, ZATTRS_JSON, ZGROUP_JSON, make_cattr
+from zarr.v3.config import RuntimeConfiguration, SyncConfiguration
 from zarr.v3.store import StoreLike, StorePath, make_store_path
-from zarr.v3.sync import sync
+from zarr.v3.sync import SyncMixin
+
+logger = logging.getLogger("zarr.group")
 
 
 @frozen
 class GroupMetadata:
     attributes: Dict[str, Any] = field(factory=dict)
-    zarr_format: Literal[3] = 3
-    node_type: Literal["group"] = "group"
+    zarr_format: Literal[2, 3] = 3  # field(default=3, validator=validators.in_([2, 3]))
+    node_type: Literal["group"] = field(default="group", init=False)
 
-    def to_bytes(self) -> bytes:
-        return json.dumps(asdict(self)).encode()
+    def to_bytes(self) -> Dict[str, bytes]:
+        if self.zarr_format == 3:
+            return {ZARR_JSON: json.dumps(asdict(self)).encode()}
+        elif self.zarr_format == 2:
+            return {
+                ZGROUP_JSON: self.zarr_format,
+                ZATTRS_JSON: json.dumps(self.attributes).encode(),
```

Comment:
how would custom JSON serialization of user attributes plug in here? e.g., i don't think numpy dtype objects serialize to JSON easily. Maybe `to_bytes` can take a `attrs_serializer` kwarg or something.